### ISSUE TITLE
Network failures with self-hosted servers

### ISSUE THREAD

**Issue by madkinsz** [PrefectHQ/prefect#7512]
This is tracking issue for various reports of network failures when self-hosting Prefect Orion.

Notably, these issues seem focused to usage of Prefect 2.6.6.
 
If adding a report to this issue, please include the following information:

- If using Prefect official Docker images for the client or server, provide the image tags
- On the server, we are interested in the Prefect version, the database, and server library versions

```
prefect version
pip freeze | grep -E '(uvicorn|starlette)'
```

- On the client, we are interested in Prefect versions and the client HTTP library versions

```
prefect version
pip freeze | grep -E '(httpx|httpcore)'
```

- Please include the **full traceback** for the error
- Check for any related error logs on the server

Related to: 
- https://github.com/PrefectHQ/prefect/issues/7472

**Comment by carlo-catalyst** [comment]
**Note**
We observed this with flows calling **run_deployment**

**Server**
- We run with docker using this image: prefecthq/prefect:2.6.6-python3.9
- Using postgres

**Agent**
- Our own custom image, with dependencies installed
- using that image, pulled these: <coming soon>



**Stack Trace**
 ```
run_deployment(name=run_deployment_name(flow=**_flow, env=env), parameters=parameters)
  File "/usr/local/lib/python3.9/site-packages/prefect/utilities/asyncutils.py", line 197, in coroutine_wrapper
    return run_async_from_worker_thread(async_fn, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/prefect/utilities/asyncutils.py", line 148, in run_async_from_worker_thread
    return anyio.from_thread.run(call)
  File "/usr/local/lib/python3.9/site-packages/anyio/from_thread.py", line 49, in run
    return asynclib.run_async_from_thread(func, *args)
  File "/usr/local/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 970, in run_async_from_thread
    return f.result()
  File "/usr/local/lib/python3.9/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/usr/local/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.9/site-packages/prefect/client/utilities.py", line 47, in with_injected_client
    return await fn(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/prefect/deployments.py", line 127, in run_deployment
    flow_run = await client.read_flow_run(flow_run_id)
  File "/usr/local/lib/python3.9/site-packages/prefect/client/orion.py", line 1439, in read_flow_run
    response = await self._client.get(f"/flow_runs/{flow_run_id}")
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1751, in get
    return await self.request(
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1527, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/usr/local/lib/python3.9/site-packages/prefect/client/base.py", line 159, in send
    await super().send(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1614, in send
    response = await self._send_handling_auth(
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1642, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1679, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "/usr/local/lib/python3.9/site-packages/httpx/_client.py", line 1716, in _send_single_request
    response = await transport.handle_async_request(request)
  File "/usr/local/lib/python3.9/site-packages/httpx/_transports/default.py", line 353, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "/usr/local/lib/python3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/local/lib/python3.9/site-packages/httpx/_transports/default.py", line 77, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.RemoteProtocolError: Server disconnected without sending a response.
```

**Comment by madkinsz** [comment]
Possibly helpful context from an httpx discussion at https://github.com/encode/httpx/discussions/2056

**Comment by iKeepo** [comment]
**Note**：
We observed this while using agent to connect the remote self host server
**Server**:
#prefect
```
Version:             2.6.5
API version:         0.8.3
Python version:      3.8.3
Git commit:          9fc2658f
Built:               Thu, Oct 27, 2022 2:24 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         hosted
```
#pip freeze | grep -E '(uvicorn|starlette)'
```
starlette==0.19.1
uvicorn==0.17.6
```
**Client**
#prefect
```
Version:             2.6.5
API version:         0.8.3
Python version:      3.8.10
Git commit:          9fc2658f
Built:               Thu, Oct 27, 2022 2:24 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         hosted
```
#pip freeze | grep -E '(httpx|httpcore)'
```
httpcore==0.15.0
httpx==0.23.0
```
**Stack Trace**:
```
Encountered exception during execution:
Traceback (most recent call last):
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_transports/default.py", line 60, in map_httpcore_exceptions
    yield
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_transports/default.py", line 353, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpcore/_async/connection_pool.py", line 253, in handle_async_request
    raise exc
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpcore/_async/connection_pool.py", line 237, in handle_async_request
    response = await connection.handle_async_request(request)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpcore/_async/connection.py", line 90, in handle_async_request
    return await self._connection.handle_async_request(request)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpcore/_async/http11.py", line 105, in handle_async_request
    raise exc
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpcore/_async/http11.py", line 84, in handle_async_request
    ) = await self._receive_response_headers(**kwargs)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpcore/_async/http11.py", line 148, in _receive_response_headers
    event = await self._receive_event(timeout=timeout)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpcore/_async/http11.py", line 191, in _receive_event
    raise RemoteProtocolError(msg)
httpcore.RemoteProtocolError: Server disconnected without sending a response.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/wind/miniconda3/lib/python3.8/site-packages/prefect/engine.py", line 580, in orchestrate_flow_run
    result = await run_sync(flow_call)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/prefect/utilities/asyncutils.py", line 68, in run_sync_in_worker_thread
    return await anyio.to_thread.run_sync(call, cancellable=True)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/home/wind/miniconda3/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/home/wind/miniconda3/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/tmp/tmpd9i_ocaqprefect/src/prefect_deployment/win_server/__init__.py", line 255, in orche_oxen_settle_fields_to_xms
    res_rsync = run_deployment(name=f"manual-handler/oxen_settlefields_{file_type}", flow_run_name=f"{func_name}_2_2")
  File "/home/wind/miniconda3/lib/python3.8/site-packages/prefect/utilities/asyncutils.py", line 197, in coroutine_wrapper
    return run_async_from_worker_thread(async_fn, *args, **kwargs)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/prefect/utilities/asyncutils.py", line 148, in run_async_from_worker_thread
    return anyio.from_thread.run(call)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/anyio/from_thread.py", line 49, in run
    return asynclib.run_async_from_thread(func, *args)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 970, in run_async_from_thread
    return f.result()
  File "/home/wind/miniconda3/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/home/wind/miniconda3/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/home/wind/miniconda3/lib/python3.8/site-packages/prefect/client/utilities.py", line 47, in with_injected_client
    return await fn(*args, **kwargs)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/prefect/deployments.py", line 127, in run_deployment
    flow_run = await client.read_flow_run(flow_run_id)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/prefect/client/orion.py", line 1432, in read_flow_run
    response = await self._client.get(f"/flow_runs/{flow_run_id}")
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_client.py", line 1751, in get
    return await self.request(
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_client.py", line 1527, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/prefect/client/base.py", line 159, in send
    await super().send(*args, **kwargs)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_client.py", line 1614, in send
    response = await self._send_handling_auth(
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_client.py", line 1642, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_client.py", line 1679, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_client.py", line 1716, in _send_single_request
    response = await transport.handle_async_request(request)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_transports/default.py", line 353, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "/home/wind/miniconda3/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/wind/miniconda3/lib/python3.8/site-packages/httpx/_transports/default.py", line 77, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.RemoteProtocolError: Server disconnected without sending a response.
```

**Comment by madkinsz** [comment]
Thanks for the report! We are currently suspicious that this is a bug in uvicorn. We'd like to find the root cause, but may explore client-side solutions in the interim.

**Comment by eudyptula** [comment]
We seen it on Prefect 2.6.3

### Server
```
Version:             2.6.3
API version:         0.8.2
Python version:      3.10.6
Git commit:          9e7da96e
Built:               Tue, Oct 18, 2022 1:55 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         <client error>
starlette==0.20.4
uvicorn==0.18.3
```

### Agent
We are running a custom build version, with our two pull requests:

- [#7361 Agent: Add limit to control number of concurrent flow runs](https://github.com/PrefectHQ/prefect/pull/7361)
- [#7362 Agent: Add tags to flow runs](https://github.com/PrefectHQ/prefect/pull/7362)

Doubt they have anything to do with it though - specially since other people are seeing the same.
```
Version:             2.6.3+3
API version:         0.8.2
Python version:      3.10.6
Git commit:          e7fd68e7
Built:               Tue, Nov 1, 2022 10:29 AM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         ephemeral
Server:
  Database:          sqlite
  SQLite version:    3.37.2
httpcore==0.15.0
httpx==0.23.0
```

### Docker container
Custom built from prefecthq/prefect:2.6.3-python3.10).
```
Version:             2.6.3
API version:         0.8.2
Python version:      3.10.8
Git commit:          9e7da96e
Built:               Tue, Oct 18, 2022 1:55 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         ephemeral
Server:
  Database:          sqlite
  SQLite version:    3.34.1
httpcore==0.15.0
httpx==0.23.0
```

### Stacktrace
```
Crash details:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 237, in handle_async_request
    response = await connection.handle_async_request(request)
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection.py", line 88, in handle_async_request
    raise ConnectionNotAvailable()
httpcore.ConnectionNotAvailable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1332, in report_task_run_crashes
    yield
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1068, in begin_task_run
    connect_error = await client.api_healthcheck()
  File "/usr/local/lib/python3.10/site-packages/prefect/client/orion.py", line 183, in api_healthcheck
    await self._client.get("/health")
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1751, in get
    return await self.request(
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1527, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/usr/local/lib/python3.10/site-packages/prefect/client/base.py", line 159, in send
    await super().send(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1614, in send
    response = await self._send_handling_auth(
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1642, in _send_handling_auth
    response = await self._send_handling_redirects(
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1679, in _send_handling_redirects
    response = await self._send_single_request(request)
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1716, in _send_single_request
    response = await transport.handle_async_request(request)
  File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 353, in handle_async_request
    resp = await self._pool.handle_async_request(req)
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 246, in handle_async_request
    async with self._pool_lock:
  File "/usr/local/lib/python3.10/site-packages/anyio/_core/_synchronization.py", line 130, in acquire
    await event.wait()
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 214, in wait
    await fut
asyncio.exceptions.CancelledError
```

Hope it helps.

**Comment by iKeepo** [comment]
@madkinsz The latest versions of uvicorn/starlette doesn't work, it still reproduces.
Never use 2.6.4 before, I'll try today.

**Comment by MuFaheemkhan** [comment]
- Prefect 2.6.6, 
- local server in docker containers. 
- self hosted postgress server in docker container
- Dask task runner

Errors:
` BrokenPipeError: [Errno 32] Broken pipe`
` ConnectionResetError: [Errno 104] Connection reset by peer`

Even without running any flows I get the above errors after some time and when I run a flow and the error appears, it crashes the flow.

Further details in this thread https://prefect-community.slack.com/archives/CL09KU1K7/p1667880546594239

**Comment by madkinsz** [comment]
@MuFaheemkhan could you please edit your post to contain the full Docker image tag if you are using one of our official images or include the versions of the libraries as requested? A full traceback for the error would also be really helpful. Thanks!

**Comment by eudyptula** [comment]
I think this is cause by this issue. Today I noticed that one of our agents had two docker containers up and running for 40 hours+. I'm pretty sure the flow run within the container crashed due to this bug, but I can't seem to find a way to relate docker container ids to the flow run information in the UI to confirm that.

It's a huge issue that prefect leaves docker containers up and running with crashed flows - eats up resources that are never released. Plus it prevents further flows in running, when using the concurrency limit I introduced in my pull request.

Currently these issues are making prefect 2 way too unstable for us to move production flows from Prefect 1.

If I'm right, this should be the log for the flow run in the GUI (same versions as mentioned above):
```
Encountered exception during execution:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/prefect/client/utilities.py", line 47, in with_injected_client
    return await fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 291, in retrieve_flow_then_begin_flow_run
    return await begin_flow_run(
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 325, in begin_flow_run
    async with AsyncExitStack() as stack:
  File "/usr/local/lib/python3.10/contextlib.py", line 714, in __aexit__
    raise exc_details[1]
  File "/usr/local/lib/python3.10/contextlib.py", line 217, in __aexit__
    await self.gen.athrow(typ, value, traceback)
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1320, in report_flow_run_crashes
    raise exc from None
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 1300, in report_flow_run_crashes
    yield
  File "/usr/local/lib/python3.10/contextlib.py", line 697, in __aexit__
    cb_suppress = await cb(*exc_details)
  File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 662, in __aexit__
    raise exceptions[0]
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 908, in create_task_run_then_submit
    task_run = await create_task_run(
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 953, in create_task_run
    task_run = await flow_run_context.client.create_task_run(
  File "/usr/local/lib/python3.10/site-packages/prefect/client/orion.py", line 1569, in create_task_run
    response = await self._client.post(
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1842, in post
    return await self.request(
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1527, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/usr/local/lib/python3.10/site-packages/prefect/client/base.py", line 182, in send
    response.raise_for_status()
  File "/usr/local/lib/python3.10/site-packages/prefect/client/base.py", line 125, in raise_for_status
    raise PrefectHTTPStatusError.from_httpx_error(exc) from exc.__cause__
prefect.exceptions.PrefectHTTPStatusError: Server error '500 Internal Server Error' for url '***'
Response: {'exception_message': 'Internal Server Error'}
For more information check: https://httpstatuses.com/500

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 287, in wrapper
    return await func(*args)
  File "/usr/local/lib/python3.10/site-packages/prefect/client/utilities.py", line 45, in with_injected_client
    async with client_context as new_client:
  File "/usr/local/lib/python3.10/site-packages/prefect/client/orion.py", line 1862, in __aexit__
    return await self._exit_stack.__aexit__(*exc_info)
  File "/usr/local/lib/python3.10/contextlib.py", line 714, in __aexit__
    raise exc_details[1]
  File "/usr/local/lib/python3.10/contextlib.py", line 697, in __aexit__
    cb_suppress = await cb(*exc_details)
  File "/usr/local/lib/python3.10/site-packages/httpx/_client.py", line 1997, in __aexit__
    await self._transport.__aexit__(exc_type, exc_value, traceback)
  File "/usr/local/lib/python3.10/site-packages/httpx/_transports/default.py", line 332, in __aexit__
    await self._pool.__aexit__(exc_type, exc_value, traceback)
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 326, in __aexit__
    await self.aclose()
  File "/usr/local/lib/python3.10/site-packages/httpcore/_async/connection_pool.py", line 312, in aclose
    raise RuntimeError(
RuntimeError: The connection pool was closed while 422 HTTP requests/responses were still in-flight.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/prefect/engine.py", line 578, in orchestrate_flow_run
    result = await run_sync(flow_call)
  File "/usr/local/lib/python3.10/site-packages/prefect/utilities/asyncutils.py", line 136, in run_sync_in_interruptible_worker_thread
    assert result is not NotSet
AssertionError
```

**Comment by MuFaheemkhan** [comment]
@eudyptula what's your prefect version? it's a different issue, I experienced the same errors as yours but got fixed after I upgraded to prefect 2.6.6
with 2.6.6 we are having the following errors:
Errors:
BrokenPipeError: [Errno 32] Broken pipe
ConnectionResetError: [Errno 104] Connection reset by peer

**Comment by madkinsz** [comment]
@eudyptula as noted by MuFaheemkhan, that is not a connection / network failure but rather a 500 error from the server. Please open a separate issue for that and include the relevant versions. There should be server logs with the error if you are hosting your own server. If you're using Cloud, we will find the server error.

> It's a huge issue that prefect leaves docker containers up and running with crashed flows - eats up resources that are never released

This should not be the case. When the flow run crashes, the process should exit. If you can reproduce this, please open an issue so we can address it.

**Comment by eudyptula** [comment]
Well, in that case we have several issues.

We were on 2.6.3, just upgraded to 2.6.4 as @madkinsz said that some users didn't experience the issue on that version. At the same time I upgraded all the dependencies of Prefect on both the agents and the server - but not on the docker containers.

Was going to see if 2.6.4 was running better before I upgraded all the way to the new version (and hopefully provide some helpful information to you at the same time). But thanks @MuFaheemkhan will definitely check out 2.6.6 and see if it helps!

@madkinsz, I'll monitor if it happens again. Are there any good ways of see output/logs from inside the docker containers? I see log output for the server and from the agent, but from the flow container I can only see the logs it posts on the Prefect UI.

**Comment by voidel** [comment]
Getting this on 2.6.7. There is no information in the traceback on why the connection failed.
Can provide additional context if required but mostly echoes what others have said.
I am using `run_deployment` and the error sporadically occurs as the subflow is running.
Error messages are:
`httpx.RemoteProtocolError: Server disconnected without sending a response`
and at other times
`httpx.ReadError` caused by `ConnectionResetError: [Errno 104] Connection reset by peer`.

To confirm, I did not have these issues on 2.6.4.

**Comment by andreas-ntonas** [comment]
I am getting the same errors.
I am running a custom Prefect 2.6.5 version (same for server and clients) with the [agent-limit feature from this PR ](https://github.com/PrefectHQ/prefect/pull/7361)(but I do not think this is related). The logs are the same as what others mention like:
`ConnectionResetError: [Errno 104] Connection reset by peer`
and
`anyio.BrokenResourceError`
and
`httpcore.ReadError`, `httpx.ReadError`

**Comment by andreas-ntonas** [comment]
In my case I managed to stop getting these network related errors by not using the **`run_deployment`** function (with None timeout)  inside tasks submitted to task runners. Instead I created my own custom task that calls directly the Orion API by using the client and functions like [`create_flow_run_from_deployment`](https://orion-docs.prefect.io/api-ref/orion/api/deployments/?h=create_flow_run_from_deployment#prefect.orion.api.deployments.create_flow_run_from_deployment) to start runs from deployments and
[`read_flow_run`](https://orion-docs.prefect.io/api-ref/orion/models/flow_runs/?h=read_flow_run#prefect.orion.models.flow_runs.read_flow_run) to regularly poll the new flow_run state

_I make sure to create a new client instance instead of getting the flow context's one in contrast to what `run_deployment` is doing when it uses @inject_client._ 

Somewhere there I believe lies the root cause of the issue: Using `run_deployment` in tasks submitted to a task_runner in parallel that also use orion's client from **flow's context** to poll the server and wait for the run to finish.

**Comment by carlo-catalyst** [comment]
We are running `run_deployment` within `tasks`, so it's possible that is having an impact.

**Comment by madkinsz** [comment]
I believe these issues are basically caused by a high volume of requests — we see these issues with the agent which polls frequently and now with `run_deploment` which also polls frequently.

**Comment by MuFaheemkhan** [comment]
In my case I stopped using run_deployment, now my flows doesn't crash, but connection reset and broken pipe error occur randomly in agent logs although this time it doesn't crash my flows.

**Comment by eudyptula** [comment]
We are seeing our `BrokenPipeError: [Errno 32] Broken pipe` in flows that are starting like 800 tasks in one go with map.
The logs also include a `RuntimeError: The connection pool was closed while 325 HTTP requests/responses were still in-flight.`

Also, the reason behind the http 500 was a database timeout, so had to increase the query timeout from the default 1 second.

So, looking from our perspective, it could very well be issues with high volumen.

**Comment by peytonrunyan** [comment]
Howdy yall! Any chance anyone here would be interested in giving this branch a shot to see if it resolves the problems? https://github.com/PrefectHQ/prefect/pull/7593#discussion_r1029528076

**Comment by voidel** [comment]
Really looking forward to a fix for this, if there's a workaround we can use to stop the parent flows failing it would be much appreciated!

**Comment by eudyptula** [comment]
@peytonrunyan: Haven't tried the branch, but is retry the right solution here? Sounds a bit like playing the lottery to me, just keep resending the same request, hoping that eventually it would go through.

If the issue is caused by high volume of requests, as @madkinsz suggests - shouldn't a solution involve some way to throttle/queue requests on the agents? (just thoughts, not that familiar with how Prefect is designed)

**Comment by madkinsz** [comment]
@eudyptula we believe this is an issue with the upstream HTTP libraries. We are implementing retries to solve the immediate issue for our users until we can work on fixing the bug upstream.

**Comment by madkinsz** [comment]
Note we believe this is addressed in [2.7.0](https://github.com/PrefectHQ/prefect/blob/main/RELEASE-NOTES.md#release-270) with #7593

**Comment by voidel** [comment]
We haven't had this error since we upgraded to 2.7.0 earlier today 👍

**Comment by colindunn** [comment]
I can also confirm that flows that were failing regularly for me with this issue now seem to be working fine

**Comment by peytonrunyan** [comment]
@madkinsz I'm going to go ahead and close this one. Feel free to reopen it if you think there's something that's not addressed.
